File: server/routes/projects.js
const express = require('express');

module.exports = (supabase) => {
    const router = express.Router();

    // GET all projects
    router.get('/', async (req, res) => {
        try {
            const { data, error } = await supabase
                .from('projects')
                .select('*')
                .order('created_at', { ascending: false });

            console.log("Fetch projects result:", data ? `Success (${data.length} records)` : "No data", "Error:", error);

            if (error) throw error;
            res.json(data);
        } catch (err) {
            res.status(500).json({ error: err.message });
        }
    });

    // POST a new project
    router.post('/', async (req, res) => {
        console.log("POST /projects request body:", req.body);
        const { title, description, location, type, budget, tags } = req.body;
        const authHeader = req.headers.authorization;

        try {
            let insertClient = supabase;
            if (authHeader) {
                insertClient = require('@supabase/supabase-js').createClient(
                    process.env.VITE_SUPABASE_URL,
                    process.env.VITE_SUPABASE_ANON_KEY,
                    { global: { headers: { Authorization: authHeader } } }
                );
            }

            const { data, error } = await insertClient
                .from('projects')
                .insert([{ title, description, location, type, budget, tags }])
                .select()
                .single();

            if (error) throw error;
            res.status(201).json(data);
        } catch (err) {
            res.status(500).json({ error: err.message });
        }
    });

    // DELETE a project
    router.delete('/:id', async (req, res) => {
        const { id } = req.params;
        const authHeader = req.headers.authorization;

        try {
            let deleteClient = supabase;
            if (authHeader) {
                deleteClient = require('@supabase/supabase-js').createClient(
                    process.env.VITE_SUPABASE_URL,
                    process.env.VITE_SUPABASE_ANON_KEY,
                    { global: { headers: { Authorization: authHeader } } }
                );
            }

            const { error } = await deleteClient
                .from('projects')
                .delete()
                .eq('id', id);

            if (error) throw error;
            res.status(204).send(); // No content
        } catch (err) {
            res.status(500).json({ error: err.message });
        }
    });

    return router;
};

File: server-python/routes/projects.py
"""
Projects API routes — Flask version.
Converted from: server/routes/projects.js
"""

from flask import Blueprint, request, jsonify
import supabase_client as db
from auth import get_token

projects_bp = Blueprint("projects", __name__)


# ---------- GET all projects ----------

@projects_bp.route("/api/projects", methods=["GET"])
def get_projects():
    """Fetch all projects, ordered by creation date (newest first)."""
    try:
        data = db.select(
            "projects",
            columns="*",
            order="created_at.desc",
        )
        
        import os
        has_service_key = os.getenv("SUPABASE_SERVICE_ROLE_KEY") is not None
        
        try:
            apps = db.select("applications", columns="project_id", use_service_key=has_service_key)
            app_counts = {}
            if apps and isinstance(apps, list):
                for app in apps:
                    pid = str(app.get("project_id"))
                    app_counts[pid] = app_counts.get(pid, 0) + 1
            
            for p in data:
                p["applicant_count"] = app_counts.get(str(p["id"]), 0)
        except Exception as e:
            print(f"Could not fetch application counts: {e}")
            for p in data:
                p["applicant_count"] = 0

        print(f"Fetch projects result: Success ({len(data)} records)")
        return jsonify(data)

    except Exception as e:
        print(f"Fetch projects error: {e}")
        return jsonify({"error": str(e)}), 500


# ---------- POST a new project ----------

@projects_bp.route("/api/projects", methods=["POST"])
def create_project():
    """Create a new project."""
    token = get_token()
    body = request.get_json()
    print(f"POST /projects request body: {body}")

    try:
        insert_data = {
            "title": body.get("title"),
            "description": body.get("description"),
            "location": body.get("location"),
            "type": body.get("type"),
            "budget": body.get("budget"),
            "hourly_rate": body.get("hourly_rate"),
            "monthly_rate": body.get("monthly_rate"),
            "deadline": body.get("deadline"),
            "tags": body.get("tags"),
        }

        results = db.insert(
            "projects",
            insert_data,
            columns="*",
            token=token,
        )

        if not results:
            return jsonify({"error": "Failed to create project"}), 500

        return jsonify(results[0]), 201

    except Exception as e:
        return jsonify({"error": str(e)}), 500


# ---------- DELETE a project ----------

@projects_bp.route("/api/projects/<project_id>", methods=["DELETE"])
def delete_project(project_id):
    """Delete a project by ID."""
    token = get_token()

    try:
        db.delete(
            "projects",
            filters={"id": project_id},
            token=token,
        )
        return "", 204

    except Exception as e:
        return jsonify({"error": str(e)}), 500
